plugins/workspace-backpack/src/index.js

```diff
@@ -438,43 +439,48 @@ export class Backpack {
    *     cleaned of all unnecessary attributes.
    */
   addItem(item) {
-    if (this.contents_.indexOf(item) !== -1) {
-      return;
-    }
+    this.addItems([item]);
+  }
 
-    this.contents_.unshift(item);
-    while (this.contents_.length > this.maxItems_) {
-      this.contents_.pop();
-    }
-    // TODO: Fire UI event for Backpack content change.
+  /**
+   * Adds multiple items to the backpack.
+   * @param {!Array<string>} items The backpack contents to add.
+   */
+  addItems(items) {
+    const filteredItems = items.filter((item) => {
+      return this.contents_.indexOf(item) === -1;
+    });
+    this.contents_.unshift(...filteredItems);
+    this.trimContents_();
+    Blockly.Events.fire(new BackpackChange(this.workspace_.id));
   }
 
   /**
    * Sets backpack contents XML.
    * @param {!Array<string>} contents The new backpack contents.
    */
   setContents(contents) {
-    this.contents_ = [...contents];
-    while (this.contents_.length > this.maxItems_) {
-      this.contents_.pop();
-    }
-    // TODO: Fire UI event for Backpack content change.
+    this.contents_ = contents.filter((item) => {
```

Comment:
Thanks for pointing it out. This is a bug. My intention was to filter the contents array passed in so that I could guarantee that all the contents in `this.contents_` were always unique (this check doesn't do that right).

Reply:
Also, looking further. I seem to have made a similar mistake in how I implemented `addItems`.

Reply:
I've refactored and added a private helper `filterDuplicates_` that handles the duplicate items correctly.